Add `#[warn(unreachable_pub)]` to a bunch of compiler crates

nnethercote opened the pull request:
By default `unreachable_pub` identifies things that need not be `pub` and tells you to make them `pub(crate)`. But sometimes those things don't need any kind of visibility. So they way I did these was to remove the visibility entirely for each thing the lint identifies, and then add `pub(crate)` back in everywhere the compiler said it was necessary. (Or occasionally `pub(super)` when context suggested that was appropriate.) Tedious, but results in more `pub` removal.

There are plenty more crates to do but this seems like enough for a first PR.

r? @compiler-errors

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9379887090/job/25825742773) [(plain)](https://github.com/rust-lang/rust/commit/fe6d4ca21542f40877914138d06fb56dc7d1b3d4/checks/25825742773/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     |
1172 |     pub fn oem_code_page() -> u32 {
     |     ---^^^^^^^^^^^^^^^^^^^^^^^^^^
     |     |
     |     help: consider restricting its visibility: `pub(crate)`
     = help: or consider exporting it for use by other crates
note: the lint level is defined here
    --> compiler/rustc_codegen_ssa/src/lib.rs:7:9
     |
     |
7    | #![deny(unreachable_pub)]
     |         ^^^^^^^^^^^^^^^

error: unreachable `pub` item
    --> compiler/rustc_codegen_ssa/src/back/link.rs:1195:5
     |
1195 |     pub fn locale_byte_str_to_string(s: &[u8], code_page: u32) -> Option<String> {
     |     ---^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |     |
     |     help: consider restricting its visibility: `pub(crate)`
     = help: or consider exporting it for use by other crates

    Checking rustc_query_impl v0.0.0 (/checkout/compiler/rustc_query_impl)
error: could not compile `rustc_codegen_ssa` (lib) due to 2 previous errors
```

</details>

Kobzol commented on compiler/rustc_ast_lowering/src/delegation.rs:
```diff
@@ -55,9 +55,9 @@ use rustc_target::spec::abi;
-    pub body_id: hir::BodyId,
-    pub sig: hir::FnSig<'hir>,
-    pub generics: &'hir hir::Generics<'hir>,
+    pub(crate) body_id: hir::BodyId,
```
Since the struct itself is `pub(crate)`, this doesn't really have any effects, right? It seems to just add more visual noise to me.

Urgau replied:
> I don't think this lint should be applied to rustc until it's changed to not warn on `pub` fields of `pub(restricted)` structs

I posted https://github.com/rust-lang/rust/pull/126040 to address this.

Urgau replied:
https://github.com/rust-lang/rust/pull/126040 was merged, the lint no longer triggers on struct fields.

Kobzol reviewed (commented):
Just my 2 cents: I prefer the style where you just use `pub` in the internal modules, and then limit visibility in the parent (for multiple reasons, one of them being that `pub(crate)` is IMO too verbose). It seems to be more uniform to me, in the sense that you export things from the module, and then decide in the parent what to export further and what not to export, rather than hard-coding the visibility to the whole crate in nested modules.

For example, in `rustc_ast_lowering/src/lib.rs`, all modules are actually private, and the crate selectively exports (or declares as `pub` in the root library module) just a few selected items. It doesn't seem to me that adding `pub(crate)` to many items in the nested modules (which are not reexported from the crate anyway) adds a lot of value.

compiler-errors commented:
Yeah, this is far more reaching than what I was asking for in the original PR (diagnostic structs). I don't have the energy to review this and also don't think we're gaining much -- this really seems like churn.

r? compiler

Nadrieril commented:
r? @Urgau since you seem interested in solving this

nnethercote commented:
I personally hate having to look at some outer level (sometimes the immediate struct, sometimes the module, sometimes a `use` item in an entirely different file) to learn that a seemingly `pub` thing is not actually `pub`. But I can see I'm in the minority.

nnethercote commented:
I also don't understand why this change would be desirable on diagnostic structs but not on other things.

Kobzol commented:
> I personally hate having to look at some outer level (sometimes the immediate struct, sometimes the module, sometimes a use item in an entirely different file) to learn that a seemingly pub thing is not actually pub.

It would be nice if IDEs could show the computed crate visibility on the struct/field. That way we could see both the final visibility, without having to specify `pub(crate)` everywhere. I'll create a ticket for RustRover :)

But otherwise, this is just a matter of taste, of course.

petrochenkov commented:
> I personally hate having to look at some outer level (sometimes the immediate struct, sometimes the module, sometimes a `use` item in an entirely different file) to learn that a seemingly `pub` thing is not actually `pub`. But I can see I'm in the minority.

I also don't like to look elsewhere, but in case of nearly all structs and many impls the outer visibility is in the immediate proximity of the fields/methods, or at worst on the same page. In those cases it's better to avoid the noise and keep just `pub`.

bors commented:
:umbrella: The latest upstream changes (presumably #126108) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
While I like this change, the concerns expressed by others makes me realize that this change is probably to big to just be approved without further discussion.

The major change proposal [page](https://forge.rust-lang.org/compiler/mcp.html#what-constitutes-a-major-change) on the forge quotes **changing coding conventions** as a potential reason for an MCP. @nnethercote if you are interested in pursuing this change further, could you open an MCP for it.

@rustbot blocked

nnethercote committed:
Add `warn(unreachable_pub)` to several crates.

It requires no additonal changes to these crates, but will prevent
unnecessary `pub`s in the future.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_arena`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_ast_lowering`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_ast_passes`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_ast_pretty`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_attr`.

nnethercote committed:
Don't add `warn(unreachable_pub)` to `rustc_baked_icu`.

But explain why.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_borrowck`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_builtin_macros`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_codegen_llvm`.

nnethercote committed:
Add `warn(unreachable_pub)` to `rustc_codegen_ssa`.

nnethercote commented:
I have updated this PR for the new `unreachable_pub` checking implemented in #126040. The diff has reduced from +946/-917 to +506/-467. Better?

Urgau commented:
I'm personally fine with the PR as is, the diff looks good to me.

@Kobzol @petrochenkov How do feel about the new diff?

Kobzol commented:
I'm still not a big fan of `pub(crate)`, but the new diff looks much better to me, so fine by me.

bors commented:
:umbrella: The latest upstream changes (presumably #127819) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
> I'm personally fine with the PR as is, the diff looks good to me.
> 
> @Kobzol @petrochenkov How do feel about the new diff?

It looks acceptable to me now.

Urgau commented:
@nnethercote could you resolve the conflicts.

@rustbot author

klensy commented on compiler/rustc_baked_icu_data/src/data/macros.rs:
```diff
@@ -18,29 +18,29 @@ macro_rules! __make_provider {
         impl $name {
             #[doc(hidden)]
             #[allow(dead_code)]
-            pub const MUST_USE_MAKE_PROVIDER_MACRO: () = ();
```
1st line contains `// @generated`, so probably this file shouldn't be changed?

Urgau replied:
Indeed, `rustc_baked_icu_data` is mostly auto-generated (except for `lib.rs`)[^1]. We should not be modifying this crate (at least not directly). I think we should just allow the lint for this crate until the generator is updated.

[^1]: https://github.com/rust-lang/rust/blob/413c9bf364f84939e05bf9d5abdccb0d6abf2fc3/compiler/rustc_baked_icu_data/src/lib.rs#L1-L4

nnethercote replied:
Fixed.

nnethercote commented:
I'm planning to file an MCP to use this lint for all (or almost all) compiler crate. This PR will serve as a good reference for that.

petrochenkov commented:
As a side note, the compiler should have enough information to automatically suggest `pub(super)` instead of `pub(crate)` when possible in `unreachable_pub` (if it doesn't already do that).

nnethercote commented:
@Urgau: the MCP has passed, so this is ready for final review. Thanks!

Urgau commented:
@bors r+ rollup

bors commented:
:pushpin: Commit cc8444274b33ffd01fd8cb35199f02697e632852 has been approved by `Urgau`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Urgau cc8444274b33ffd01fd8cb35199f02697e632852 -->
<!-- homu: {"type":"Approved","sha":"cc8444274b33ffd01fd8cb35199f02697e632852","approver":"Urgau","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
